src/inputs/TreeSelectField/TreeSelectField.tsx

```diff
@@ -157,29 +164,53 @@ function TreeSelectFieldBase<O, V extends Value>(props: TreeSelectFieldProps<O,
-        : []),
-    ];
-  }
+  const levelOptions = useCallback(
+    (o: NestedOption<O>, level: number, filtering?: boolean): LeveledOption<O>[] => {
+      // If a user is filtering, then do not provide level to the options as the various paddings may look quite odd.
+      const actualLevel = filtering ? 0 : level;
+      return [
+        [o, actualLevel],
+        ...(o.children?.length && !collapsedKeys.includes(valueToKey(getOptionValue(o)))
+          ? o.children.flatMap((oc: NestedOption<O>) => levelOptions(oc, actualLevel + 1, filtering))
+          : []),
+      ];
+    },
+    [collapsedKeys, getOptionValue],
+  );
 
-  // Initialize the TreeFieldState
-  const [fieldState, setFieldState] = useState<TreeFieldState<O>>(() => {
+  // Given an option and the selected keys, it will return the top most selected option in the option's tree.
+  // This could be a parent of the option, or the option itself.
+  const getTopLevelSelections = useCallback(
+    (o: NestedOption<O>, selectedKeys: Set<Key>): NestedOption<O>[] => {
+      // If this element is already selected, return early. Do not bother looking through children.
+      if (selectedKeys.has(valueToKey(getOptionValue(o)))) return [o];
+      // If this element has no children, then look through the children for top level selected options.
+      if (o.children) return [...o.children.flatMap((c) => getTopLevelSelections(c, selectedKeys))];
+      return [];
+    },
+    [getOptionValue],
+  );
+
+  // `disabledKeys` from ComboBoxState does not support additional meta for showing a disabled reason to the user
+  // This lookup map helps us cleanly prune out the optional reason text, then access it further down the component tree
+  const disabledOptionsWithReasons = Object.fromEntries(disabledOptions?.map(disabledOptionToKeyedTuple) ?? []);
+
+  const initTreeFieldState = useCallback(() => {
     const filteredOptions = initialOptions.flatMap((o) => levelOptions(o, 0));
-    const selectedOptions: NestedOption<O>[] =
+    const maybeDuplicatedSelectedOptions: NestedOption<O>[] =
       values?.flatMap((v) => {
         const maybeOption = findOption(initialOptions, valueToKey(v), getOptionValue);
-        if (!maybeOption) return [];
-
-        const { option } = maybeOption;
-        // If the selected key has children then all children should also be considered selected.
-        return [option, ...(option.children?.flatMap(flattenOptions) ?? [])];
+        if (maybeOption.length === 0) return [];
+        const options: NestedOption<O>[] = [];
+        for (const { option } of maybeOption) {
+          options.push(option, ...(option.children?.flatMap(flattenOptions) ?? []));
+        }
+        return options;
       }) ?? [];
+    // Dedupe selectedOptions
+    const selectedOptions: NestedOption<O>[] = maybeDuplicatedSelectedOptions.filter(
+      (o, idx, self) => idx === self.findIndex((t) => getOptionValue(o) === getOptionValue(t)),
```

Comment:
Thought: this is technically an O(n^2) dedup, which if we get to ~1000 elements in the list will be slow.

Just curious, but what's the use case for letting clients send in dup values, and wanting TreeSelectField to be responsible for deduping them?

Reply:
Yea, I need to look at this closer. Admittedly I was throwing this together as quick as possible to help unblock the digital orchestration team.

Reply:
The user case for letting clients send in duplicated values... tbh I'm not sure. Good push to revisit the need for this here.

Reply:
I ended up refactoring this. I believe the set up should be much better now.